packages/commonwealth/server/workers/sitemap/createDatabasePaginator.ts

```diff
@@ -0,0 +1,204 @@
+ */
+export interface Page {
+  readonly links: ReadonlyArray<Link>;
+}
+
+/**
+ * Interface representing a Paginator.
+ *
+ * A Paginator is used to navigate through a collection of pages.  This way
+ * we can avoid fetching all records into memory.
+ */
+export interface Paginator {
+  readonly hasNext: () => Promise<boolean>;
+  readonly next: () => Promise<Page>;
+}
+
+interface TableAdapter {
+  toRecord: (obj: object) => Link;
+  executeQuery: (ptr: number, limit: number) => Promise<ReadonlyArray<object>>;
+}
+
+export function createDatabasePaginatorDefault(limit: number = 50000) {
+  const threads = createDatabasePaginatorWithAdapter(
+    createThreadsTableAdapter(),
+    limit,
+  );
+  const profiles = createDatabasePaginatorWithAdapter(
+    createProfilesTableAdapter(),
+    limit,
+  );
+
+  return { threads, profiles };
+}
+
+function createThreadsTableAdapter(): TableAdapter {
+  type ThreadInstancePartial = Pick<
+    ThreadInstance,
+    'id' | 'updated_at' | 'title' | 'community_id'
+  >;
+
+  function toRecord(obj: object) {
+    const thread = obj as ThreadInstancePartial;
+    const url = getThreadUrl({
+      chain: thread.community_id,
+      id: thread.id,
+      title: thread.title,
+    });
+    return {
+      id: thread.id,
+      url,
+      updated_at: thread.updated_at.toISOString(),
+    };
+  }
+
+  async function executeQuery(
+    ptr: number,
+    limit: number,
+  ): Promise<ReadonlyArray<object>> {
+    return await models.sequelize.query(
+      `
+          SELECT "Threads".id,
+                 "Threads".updated_at,
+                 "Threads".title,
+                 "Threads".community_id
+          FROM "Threads"
+          WHERE "Threads".id > ${ptr}
+          ORDER BY "Threads".id
+          LIMIT ${limit};
+      `,
+      { type: QueryTypes.SELECT },
+    );
+  }
+  return { toRecord, executeQuery };
+}
+
+function createProfilesTableAdapter(): TableAdapter {
+  type ProfileInstancePartial = Required<Pick<ProfileAttributes, 'id'>> &
+    Pick<ProfileAttributes, 'updated_at'>;
+
+  function toRecord(obj: object) {
+    const profile = obj as ProfileInstancePartial;
+    const url = `https://commonwealth.im/profile/id/${profile.id}`;
+    return {
+      id: profile.id,
+      url,
+      updated_at: profile.updated_at.toISOString(),
+    };
+  }
+
+  async function executeQuery(
+    ptr: number,
+    limit: number,
+  ): Promise<ReadonlyArray<object>> {
+    return await models.sequelize.query(
+      `
+          SELECT "Profiles".id,
+                 "Profiles".updated_at
+          FROM "Profiles"
+          WHERE "Profiles".id > ${ptr}
+          ORDER BY "Profiles".id
+          LIMIT ${limit};
+      `,
+      { type: QueryTypes.SELECT },
+    );
+  }
+  return { toRecord, executeQuery };
+}
+
+export function createDatabasePaginatorWithAdapter(
+  adapter: TableAdapter,
+  limit: number,
+): Paginator {
+  // the page we're on...
+  let idx = 0;
+
+  let ptr = -1;
```

Comment:
in case this process needs to be stopped (or crashes)... should we store watermarks (options: db, filesystem, s3) we it can recover from the last page?

Reply:
@timolegros and I are talking about the scale issues now. 

I think this is a valid point but the entire dump should take very little time.  It also only has to be done once every day really or maybe once an hour.

I still don't know how long it's going to take but this is as efficient as it's going to get for a sequential read.

There are some other strategies we can take but they are all going to be somewhat complicated.

Reply:
We could use watermarks but it's probably a simpler implementation to just overwrite (start over from 0). Technically, we could also base future updates on the Outbox and just replay creation/deletion events.

Reply:
that also works, a "backfill job" that runs daily/weekly/whatever depending on size... and a real time policy that updates from events

Reply:
If it was just 'append only' it would be easy but the main issue are deletes and updates. The updated_at is changed an then I change the valid in sitemaps so it can be re-crawled.  

Once that changes I have to rewrite everything anyway.

It's a longer conversation but this should scale for us for 6-12 months. 

If it STOPS scaling that's a good thing because it means we're crushing it in terms of growth.